### ISSUE TITLE
[bug] `gobble` results in malformed formatting, `autogobble` works

### ISSUE THREAD

**Issue by goyalyashpal** [gpoore/minted#379]
![image](https://github.com/gpoore/minted/assets/19423063/f22b2ed5-9c97-4bf5-a44b-78215c2d00bb)

### MWE:

Note: here 
* the 4 spaces are caused due to indentation inside the `filecontents` block,
* so, `autogobble` and `gobble=4` shall be equivalent, 
* as can be seen from eyeballing too, that both the outputs align
* but something else too is happening in gobble (like say maybe insertion of some ghost character or change of encoding etc) which causes the pygments to fail to recognize it as proper file & hence results in malformed formatting
* compiled with `xelatex --shell-escape document.tex`

```latex
\begin{filecontents}[noheader, overwrite]{./test.mysql}
    mysql-> select * from firstLine;
    mysql-> select * from secondLine;
    mysql-> \help contents
    mysql-> \h contents
\end{filecontents}

\documentclass{article}
\usepackage{minted}


\begin{document}

\inputminted[autogobble]{psql}{test.mysql}    % Works
\inputminted[gobble=4]{psql}{test.mysql}    % Buggy

\end{document}

```

----
just for my reference, first noticed `filecontents` at https://github.com/gpoore/minted/issues/378#issuecomment-1794025908

**Comment by muzimuzhi** [comment]
Seems a [`pygments`](https://github.com/pygments/pygments) bug.

```bash
$ cat test.mysql
    mysql-> select * from firstLine;
    mysql-> select * from secondLine;
    mysql-> \help contents
    mysql-> \h contents

$ cat test_noindent.mysql
mysql-> select * from firstLine;
mysql-> select * from secondLine;
mysql-> \help contents
mysql-> \h contents

# see outputs in screenshot below
$ pygmentize -l psql -F gobble:n=4 test.mysql

$ pygmentize -l psql test_noindent.mysql
```

![image](https://github.com/gpoore/minted/assets/6376638/89bf7ad9-21df-4799-b9ab-086ed6092018)

**Comment by muzimuzhi** [comment]
`minted` option `gobble=4` is passed to `pygments` executable `pygmentize` as CLI flag `-F gobble:n='4'`. But unfortunately, `pytments` "filters are applied _after_ lexing" (from https://github.com/pygments/pygments/issues/585#issuecomment-526848356).

Also see `pygments` [doc on filters](https://pygments.org/docs/filters/)

> An arbitrary number of filters can be applied to token streams coming from lexers to improve or annotate the output.

Hence if an indented PostgreSQL console (psql) should be correctly tokenized, it's [`PostgresConsoleLexer`](https://github.com/pygments/pygments/blob/33c66e714e35b345ac634691488fac564589ced6/pygments/lexers/sql.py#L302) to blame.

**Comment by goyalyashpal** [comment]
hi! thanks a lot for such thorough investigation..


> if an indented PostgreSQL console (psql) should be correctly tokenized, it's [`PostgresConsoleLexer`](https://github.com/pygments/pygments/blob/33c66e714e35b345ac634691488fac564589ced6/pygments/lexers/sql.py#L302) to blame.

ehm, no. 'cz the console input-output is never indented i.e. false at the "if" itself. so, the pgconsole lexer is perfectly fine the way it works.


> `minted` option `gobble=4` is passed to `pygments` executable `pygmentize`

if minted can autogobble, then can it not also gobble?


> But unfortunately, `pytments` "filters are applied _after_ lexing" (from [pygments/pygments#585 (comment)](https://github.com/pygments/pygments/issues/585#issuecomment-526848356)).

the issue lies here only.  


regardless of where the solution is supposed to come; i think this behaviour of _"gobble as a filter to pygments which's then applied there after lexing"_ is worth a place in documentation.

**Comment by goyalyashpal** [comment]
regardless of where the solution is supposed to come; i think this behaviour is worth a place in documentation:
* _"gobble as a filter to pygments which's then applied there after lexing"_ , or more clearly & succinctly as:
* _"autogobble:_ a minted's preprocessing option, _gobble:_ a pygments' post-lexing filter"

**Comment by gpoore** [comment]
`gobble` and `autogobble` behave differently because only `gobble` is a Pygments filter. All of this will change in `minted` v3.0, since there will be a `minted` executable where filter-type behavior can be implemented separately from Pygments, both before and after lexing. At that point, all gobbling can be implemented in a uniform fashion. I'm about to release a new LaTeX data serialization package that will be the foundation of much of `minted` v3.0, and then will be able to start working on the new executable.

**Comment by gpoore** [comment]
This is fixed in `minted` v3.0, which is available as a [beta on GitHub](https://github.com/gpoore/minted/tree/v3-dev) and will have a final release soon. `gobble` now operates before syntax highlighting, just like `autogobble`. There is a new option `gobblefilter` that provides the old `gobble` functionality using the Pygments gobble filter.
